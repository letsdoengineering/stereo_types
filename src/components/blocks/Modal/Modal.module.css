/* Centre modal in view port */
.modal {
    position: fixed;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
    min-width: 45rem;
    min-height: 20rem;
    display: none;
    padding: 16px;
    z-index: 1;
}

/* Apply standard focus style */
.modal:focus {
    outline: 5px auto Highlight;
    outline: 5px auto -webkit-focus-ring-color;
}

/* Set an opaque cover over page content behind the modal, keeps mouse clicks on modal where ever they are in the page */
.modal:before {
    content: "";
    position: fixed;
    top: -10000px;
    left: -5000px;
    width: 10000px;
    height: 90000px;
    background: #33333340;
    z-index: -2;
}

 /* To give modal white background, needs to be applied to :after which is above the :before. (background style on .modal is below the :before) */
.modal:after {
    content: "";
    position: fixed;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    background-color: var(--white);
    z-index: -1;
}

/* Modal is only visible when open; default display = none. */
.isOpen {
    display: block;
}

/* add invisible and v. quick transition so we can listen to focus leaving modal */
.modal:not(:focus-within) {
    transition: background-color 0.01s;
    background-color: rgb(255, 255, 254);
}

.header {
    margin-right: 40px;
    margin-bottom: 8px;
    display: flex;
    align-items: center;
}

.header:focus {
    outline: 5px auto Highlight;
    outline: 5px auto -webkit-focus-ring-color;
}

/* close button is the last element in the dialog (for cycling tab back to modal), but needs visually placing in the top right for display */
.closeButton {
    position: fixed;
    top: 16px;
    right: 24px;
}
