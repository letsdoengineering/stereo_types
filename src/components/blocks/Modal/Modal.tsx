import React, { useRef, useEffect, useState } from 'react'
import getClassNames from '../../../utils/get-class-names'

import Card from '../../blocks/Card/Card'
import IconButton from '../../basics/IconButton/IconButton'
import * as self from './Modal'

import modalStyles from './Modal.module.css'

export type ModalProps = {
  /** boolean for setting modal open */
  isOpen: boolean
  /** callback for when its closed - triggered by escape key, close button or mouse click outside the dialog  */
  setClosed(): void
  /** element id that should take focus when closing modal, normally the button which opened it */
  returnFocusId: string
  /** classname for applying custom styles */
  className?: string
  /** header content */
  headerContent: React.ReactNode
  /** main content */
  children?: React.ReactNode
}

/** isEventOutsideBoundary: a function that takes a mouse click event and DOMRect, then returns true if the click x and y is deemed outside the dom rectangle points - all relative to view port.
 @param { MouseEvent }  event - a mouse click event
 @param { DOMRect } boundaryRect - DOMRect obtained from calling .getBoundingClientRect() on a element in the dom
 */
export const isEventOutsideBoundary = (event: React.MouseEvent, boundaryRect: DOMRect): boolean => {
  return (
    boundaryRect?.left > event.clientX ||
    boundaryRect?.right < event.clientX ||
    boundaryRect?.top > event.clientY ||
    boundaryRect?.bottom < event.clientY
  )
}

/** Modal: Renders a dialog when isOpen = true, displaying a close button and the header and children as content.
 * On opening it takes focus and hides any content outside the viewport to prevent scrolling
 * When escape key or close button are pressed, or mouse is clicked outside the box it calls setClosed and un-hides overflow content to enable scrolling
 * */
const Modal: React.FC<ModalProps> = ({
  children,
  className,
  headerContent,
  isOpen,
  setClosed,
  returnFocusId,
  ...rest
}: ModalProps) => {
  const modalClassNames = getClassNames({
    defaultClasses: [modalStyles.modal],
    className: className,
    conditionalClasses: {
      [modalStyles.isOpen]: isOpen,
    },
  })
  const [modalBoundaries, setModalBoundaries] = useState<DOMRect>()
  const ESC_KEY = 'Escape'
  const TAB_KEY = 'Tab'
  const closeButton = useRef<HTMLButtonElement>(null) // to set focus when tabbing off the page
  const modalDiv = useRef<HTMLDivElement>(null)

  useEffect(() => {
    if (isOpen && modalDiv.current) {
      modalDiv.current.focus() // On opening set focus on modal
      setModalBoundaries(modalDiv.current.getBoundingClientRect()) // Fetch modal x/y boundaries
      modalDiv.current.addEventListener(
        'transitionend',
        () => modalDiv.current?.focus() // Listen for css transition caused by focus leaving dialog and send focus back to modal
      )
    }
  }, [isOpen])

  const handleClose = (): void => {
    document.body.style.overflow = 'visible' // Enable scrolling on the page again
    setClosed() // CallBack for when close triggered; parent controls "isOpen" state.
    const returnFocusTo = document.getElementById(returnFocusId) // Get element who's Id was passed to return focus
    returnFocusTo?.focus() // Set focus back to element (button) that lost focus to modal
  }

  const handleModalKeyDown = (e: React.KeyboardEvent): void => {
    // Close when escape key pressed
    if (e.key === ESC_KEY) {
      handleClose()
    }
  }

  const handleCloseButtonKeyDown = (e: React.KeyboardEvent): void => {
    // Set focus on modal if tabbing while focused on close button
    if (e.key === TAB_KEY) {
      e.preventDefault()
      modalDiv.current?.focus()
    }
  }

  const handleMouseDown = (e: React.MouseEvent): void => {
    if (modalBoundaries && self.isEventOutsideBoundary(e, modalBoundaries)) handleClose()
  }

  // Do not render anything if closed
  if (!isOpen) return null
  else {
    document.body.style.overflow = 'hidden' // Prevent scrolling the page
    return (
      <Card
        className={modalClassNames}
        role='dialog'
        aria-labelledby='modal-description'
        aria-modal='true'
        onMouseDown={(e: React.MouseEvent): void => handleMouseDown(e)}
        onKeyDown={handleModalKeyDown}
        tabIndex={0}
        ref={modalDiv}
        {...rest}
      >
        <div className={modalStyles.header}>{headerContent}</div>
        {children}
        <IconButton
          buttonClassName={modalStyles.closeButton}
          icon='times'
          label='close modal'
          onClick={handleClose}
          onKeyDown={handleCloseButtonKeyDown}
          ref={closeButton}
        />
        <div className='visually-hidden' id='modal-description'>
          This dialog window is overlaying the page content. The last element is the close button
          which will take you back where you were before.
        </div>
      </Card>
    )
  }
}

export default Modal
