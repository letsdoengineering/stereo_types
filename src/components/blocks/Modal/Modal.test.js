import * as React from 'react'
import renderer from 'react-test-renderer'
import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { axe } from 'jest-axe'

import * as allModal from './Modal'
const { default: Modal } = allModal

describe('Modal', () => {
    afterEach(() => {
        jest.clearAllMocks()
        jest.restoreAllMocks()
    })
    it('calls setClosed when close button is clicked', () => {
        const mockSetClosed = jest.fn()
        render(
            <Modal
                isOpen={true}
                returnFocusId={'a unique id'}
                setClosed={mockSetClosed}
                headerContent={<h2>A modal box</h2>}
            >
                <p>content</p>
            </Modal>
        )
        const closeButton = screen.getByText('close modal')
        userEvent.click(closeButton)
        expect(mockSetClosed).toHaveBeenCalled()
    })

    it('calls setClosed when escape key is pressed', () => {
        const mockSetClosed = jest.fn()
        render(
            <Modal
                isOpen={true}
                returnFocusId='a-unique-id'
                setClosed={mockSetClosed}
                headerContent={<h2>A modal box</h2>}
            >
                <p>content</p>
            </Modal>
        )
        userEvent.type(screen.getByText('A modal box'), '{esc}')
        expect(mockSetClosed).toHaveBeenCalled()
    })

    it('calls isEventOutsideBoundary when mouse is clicked, and if TRUE is returned it also calls setClosed call back', () => {
        const mockEventSpy = jest
            .spyOn(allModal, 'isEventOutsideBoundary')
            .mockReturnValue(true)

        const mockSetClosed = jest.fn()
        render(
            <div>
                <Modal
                    isOpen={true}
                    returnFocusId='a-unique-id'
                    setClosed={mockSetClosed}
                    headerContent={<h2>A modal box</h2>}
                >
                    <p>content</p>
                </Modal>
            </div>
        )

        userEvent.click(screen.getByText('content'))
        expect(mockEventSpy).toHaveBeenCalled()

        // should also call setClosed since mock return value of isEventOutsideBoundary = true
        expect(mockSetClosed).toHaveBeenCalled()
    })

    it('calls isEventOutsideBoundary when mouse is clicked, and if FALSE is returned it does not call setClosed call back', () => {
        const mockEventSpy = jest
            .spyOn(allModal, 'isEventOutsideBoundary')
            .mockReturnValue(false)

        const mockSetClosed = jest.fn()
        render(
            <div>
                <Modal
                    isOpen={true}
                    returnFocusId='a-unique-id'
                    setClosed={mockSetClosed}
                    headerContent={<h2>A modal box</h2>}
                >
                    <p>content</p>
                </Modal>
            </div>
        )

        userEvent.click(screen.getByText('content'))
        expect(mockEventSpy).toHaveBeenCalled()

        // should not call setClosed since mock return value of isEventOutsideBoundary = false
        expect(mockSetClosed).not.toHaveBeenCalled()
    })

    it('focus is set to modal when opened', () => {
        const mockSetClosed = jest.fn()
        render(
            <div>
                <button>outside before</button>
                <Modal
                    isOpen={true}
                    returnFocusId='a-unique-id'
                    setClosed={mockSetClosed}
                    headerContent={<h2>A modal box</h2>}
                >
                    <p>content</p>
                </Modal>
                <button>outside after</button>
            </div>
        )
        const modal = screen.getByRole('dialog')
        expect(modal).toHaveFocus()
    })

    it('when close button has focus, pressing tab key redirects focus back to the modal', () => {
        const mockSetClosed = jest.fn()
        render(
            <div>
                <button>outside before</button>
                <Modal
                    isOpen={true}
                    returnFocusId='a-unique-id'
                    setClosed={mockSetClosed}
                    headerContent={<h2>A modal box</h2>}
                >
                    <p>content</p>
                </Modal>
                <button>outside after</button>
            </div>
        )
        const closeButton = screen.getByText('close modal').closest('button')
        closeButton.focus()
        expect(closeButton).toHaveFocus()

        userEvent.tab()
        const modal = screen.getByRole('dialog')
        expect(modal).toHaveFocus()
    })

    it('when close button has focus, pressing keys other than tab key does not redirect focus back to the modal', () => {
        const mockSetClosed = jest.fn()
        render(
            <div>
                <button>outside before</button>
                <Modal
                    isOpen={true}
                    returnFocusId='a-unique-id'
                    setClosed={mockSetClosed}
                    headerContent={<h2>A modal box</h2>}
                >
                    <p>content</p>
                </Modal>
                <button>outside after</button>
            </div>
        )
        const closeButton = screen.getByText('close modal').closest('button')
        closeButton.focus()
        expect(closeButton).toHaveFocus()

        userEvent.type(closeButton, 'p')
        const modal = screen.getByRole('dialog')
        expect(modal).not.toHaveFocus()
    })

    it('has event listener css transition (caused by not:focusWithin) which will redirect focus back to the modal if it leaves', () => {
        const mockSetClosed = jest.fn()
        render(
            <div>
                <button>outside before</button>
                <Modal
                    isOpen={true}
                    returnFocusId='a-unique-id'
                    setClosed={mockSetClosed}
                    headerContent={<h2>A modal box</h2>}
                >
                    <p>content</p>
                </Modal>
                <button>outside after</button>
            </div>
        )
        const modal = screen.getByRole('dialog')
        expect(modal).toHaveFocus()

        const outsideModalButton = screen.getByText('outside after')
        outsideModalButton.focus()
        expect(outsideModalButton).toHaveFocus()

        let event = document.createEvent('Event')
        event.initEvent('transitionend', true, true)
        modal.dispatchEvent(event)
        expect(modal).toHaveFocus()
    })

    it('renders nothing when isOpen = false', () => {
        const mockSetClosed = jest.fn()
        render(
            <Modal
                isOpen={false}
                returnFocusId='a-unique-id'
                setClosed={mockSetClosed}
                headerContent={<h2>A modal box</h2>}
            >
                <p>content</p>
            </Modal>
        )
        const closeButton = screen.queryByText('close modal')
        expect(closeButton).not.toBeInTheDocument()
    })

    it('has not changed the JSX layout of the page from previous snapshot', () => {
        const mockSetClosed = jest.fn()
        const tree = renderer
            .create(
                <Modal
                    isOpen={true}
                    returnFocusId='a-unique-id'
                    setClosed={mockSetClosed}
                    headerContent={<h2>A modal box</h2>}
                >
                    <p>content</p>
                </Modal>
            )
            .toJSON()

        expect(tree).toMatchSnapshot()
    })

    it('has not changed the JSX layout of the page from previous snapshot - closed', () => {
        const mockSetClosed = jest.fn()
        const tree = renderer
            .create(
                <Modal
                    isOpen={false}
                    returnFocusId='a-unique-id'
                    setClosed={mockSetClosed}
                    headerContent={<h2>A modal box</h2>}
                >
                    <p>content</p>
                </Modal>
            )
            .toJSON()

        expect(tree).toMatchSnapshot()
    })

    it('renders the page DOM that is accessible', async () => {
        const mockSetClosed = jest.fn()
        const { container } = render(
            <Modal
                returnFocusId='a-unique-id'
                isOpen={true}
                setClosed={mockSetClosed}
                headerContent={<h2>A modal box</h2>}
            >
                <p>content</p>
            </Modal>
        )
        const results = await axe(container)

        expect(results).toHaveNoViolations()
    })
})

describe('isEventOutsideBoundary', () => {
    it('returns false when event coordinates are within the rectangle boundaries', () => {
        const mockMouseEvent = { clientX: 15, clientY: 15 }
        const mockDOMRect = { left: 10, right: 20, top: 10, bottom: 20 }
        expect(
            allModal.isEventOutsideBoundary(mockMouseEvent, mockDOMRect)
        ).toEqual(false)
    })

    it('returns true when event coordinates are within the rectangle boundaries', () => {
        const mockMouseEvent = { clientX: 1, clientY: 1 }
        const mockDOMRect = { left: 10, right: 20, top: 10, bottom: 20 }
        expect(
            allModal.isEventOutsideBoundary(mockMouseEvent, mockDOMRect)
        ).toEqual(true)
    })
})
