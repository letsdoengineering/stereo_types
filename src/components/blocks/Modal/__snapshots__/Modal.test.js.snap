// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal has not changed the JSX layout of the page from previous snapshot - closed 1`] = `null`;

exports[`Modal has not changed the JSX layout of the page from previous snapshot 1`] = `
<div
  aria-labelledby="modal-description"
  aria-modal="true"
  className="card modal isOpen"
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  role="dialog"
  tabIndex={0}
>
  <div
    className="header"
  >
    <h2>
      A modal box
    </h2>
  </div>
  <p>
    content
  </p>
  <button
    className="button closeButton"
    disabled={false}
    onClick={[Function]}
    onKeyDown={[Function]}
    type="button"
  >
    <span
      aria-hidden="true"
      className="fas fa-times iconSizeL iconColorGrey buttonIcon"
    />
    <span
      className="visually-hidden"
    >
      close modal
    </span>
  </button>
  <div
    className="visually-hidden"
    id="modal-description"
  >
    This dialog window is overlaying the page content. The last element is the close button which will take you back where you were before.
  </div>
</div>
`;
